### PR TITLE
nn.Normalize

### PR DESCRIPTION
**Pull request by Atticus1806**
#188 
adds `Normalize` and documents `normalize`

Note that 
```python
    if isinstance(axis, nn.Dim):
      axis = [axis]
``` 
is because `Parameter` does not support `nn.Dim`, but only `Sequence[nn.Dim]`, maybe we can also just catch this in `Parameter`.

### CODE REVIEW

File: nn/normalization.py [hunk 1]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+    if isinstance(axis, nn.Dim):
+      axis = [axis]
+    self.scale = nn.Parameter(shape=axis, trainable=True) if scale else None
+    self.bias = nn.Parameter(shape=axis, trainable=True) if bias else None
+
+  def __call__(self, a: nn.Tensor):
+    """
+
+    :param a: input
+    :return:
```
Comment (albertz) [inline nn/normalization.py:217]: Such docstring does not really add any information. Just leave it away. Shorter code can be read faster and thus is easier to understand.

File: nn/normalization.py [hunk 2]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+    """
+
+    norm = normalize(a, axis=self.axis, epsilon=self.epsilon)
```
Comment (albertz) [inline nn/normalization.py:220]: I don't think it makes sense to use `self.axis` here.

File: nn/normalization.py [hunk 3]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
 
   mean, variance = nn.moments(a, axis=axis)
   return (a - mean) * nn.rsqrt(variance + epsilon)
+
+
+class Normalize(nn.Module):
+  """
+  Wrapper around normalize with additional scale and bias
```
Comment (albertz) [inline nn/normalization.py:191]: You should refer to the function correctly.
```suggestion
  Wrapper around :func:`normalize` with additional scale and bias
```

Comment (albertz) [reply nn/normalization.py:191]: But also, I think describing this as "wrapper" is confusing.
Just say:
```suggestion
  :func:`normalize` with additional scale and bias
```

File: nn/normalization.py [hunk 4]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+    :param epsilon: epsilon for numerical stability
+    :param scale: whether to include a trainable scale
+    :param bias: whether to include a trainable bias
+    """
+    super(Normalize, self).__init__()
+    self.epsilon = epsilon
+    self.axis = axis
+    if isinstance(axis, nn.Dim):
+      axis = [axis]
+    self.scale = nn.Parameter(shape=axis, trainable=True) if scale else None
```
Comment (albertz) [inline nn/normalization.py:210]: You don't need the `trainable` argument. The code should be consistent to other code.

File: nn/normalization.py [hunk 5]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+
+  def __call__(self, a: nn.Tensor):
+    """
+
+    :param a: input
+    :return:
+    """
+
+    norm = normalize(a, axis=self.axis, epsilon=self.epsilon)
+    if self.scale:
```
Comment (albertz) [inline nn/normalization.py:221]: I think you should better check for `is not None`:
```suggestion
    if self.scale is not None:
```

File: nn/normalization.py [hunk 6]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+      norm = self.scale * norm
+    if self.bias:
```
Comment (albertz) [inline nn/normalization.py:223]:
```suggestion
    if self.bias is not None:
```

File: nn/normalization.py [hunk 7]
```diff
@@ -183,3 +184,44 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+      norm = norm + self.bias
+
+    return norm
+
```
Comment (albertz) [inline nn/normalization.py:227]: See failing test.
```
returnn_common/nn/normalization.py:227: WEAK WARNING Pep8CodeStyle: W391: blank line at end of file
```
```suggestion
```

File: nn/normalization.py (hunk 4 above)
Comment (Atticus1806) [reply nn/normalization.py:210]: Why do I not need it? Its default is not `True`.

Comment (albertz) [reply nn/normalization.py:210]: It is (on RETURNN side). And in any case, it should be consistent to other code.

Comment (Atticus1806) [reply nn/normalization.py]: Ah okay, I did not know that.

File: nn/normalization.py [hunk 8]
```diff
@@ -183,3 +184,35 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+    :param scale: whether to include a trainable scale
+    :param bias: whether to include a trainable bias
+    """
+    super(Normalize, self).__init__()
+    self.epsilon = epsilon
+    self.scale = nn.Parameter(shape=param_shape) if scale else None
+    self.bias = nn.Parameter(shape=param_shape) if bias else None
+
+  def __call__(self, a: nn.Tensor, axis: Union[nn.Dim, Sequence[nn.Dim]]):
+
```
Comment (albertz) [inline nn/normalization.py:211]: This empty line is inconsistent to our other code.

File: nn/normalization.py [hunk 9]
```diff
@@ -183,3 +184,35 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+    norm = normalize(a, axis=axis, epsilon=self.epsilon)
+    if self.scale is not None:
+      norm = self.scale * norm
+    if self.bias is not None:
+      norm = norm + self.bias
+
```
Comment (albertz) [inline nn/normalization.py:217]: I don't really see how this empty line adds readability.

File: nn/normalization.py [hunk 10]
```diff
@@ -183,3 +184,35 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
+  """
+  :func:`normalize` with additional scale and bias
+  """
+
+  def __init__(self,
+               param_shape: Sequence[nn.Dim],
+               epsilon: float = 1e-6,
+               scale: bool = True, bias: bool = True):
+    """
+
```
Comment (albertz) [inline nn/normalization.py:199]: There should be no empty line here (unless you have some initial description, which you do not have here).

File: nn/normalization.py [hunk 11]
```diff
@@ -183,3 +184,35 @@ def normalize(a: nn.Tensor, *, axis: Union[nn.Dim, Sequence[nn.Dim]], epsilon: f
...
   mean, variance = nn.moments(a, axis=axis)
   return (a - mean) * nn.rsqrt(variance + epsilon)
+
+
+class Normalize(nn.Module):
+  """
+  :func:`normalize` with additional scale and bias
+  """
+
+  def __init__(self,
```
Comment (albertz) [inline nn/normalization.py:194]: I think all the args should be named args.
```suggestion
  def __init__(self, *,
```

**Review by albertz — APPROVED** [review]
